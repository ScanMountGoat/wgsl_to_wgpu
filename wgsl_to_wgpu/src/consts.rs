use proc_macro2::{Span, TokenStream};
use quote::quote;
use syn::Ident;

pub fn consts(module: &naga::Module) -> Vec<TokenStream> {
    // Create matching Rust constants for WGSl constants.
    module
        .constants
        .iter()
        .filter_map(|(_, t)| -> Option<TokenStream> {
            let name = Ident::new(t.name.as_ref()?, Span::call_site());

            // TODO: Add support for f64 and f16 once naga supports them.
            let type_and_value = match &module.global_expressions[t.init] {
                naga::Expression::Literal(literal) => match literal {
                    naga::Literal::F64(v) => Some(quote!(f32 = #v)),
                    naga::Literal::F32(v) => Some(quote!(f32 = #v)),
                    naga::Literal::U32(v) => Some(quote!(u32 = #v)),
                    naga::Literal::I32(v) => Some(quote!(i32 = #v)),
                    naga::Literal::U64(v) => Some(quote!(u64 = #v)),
                    naga::Literal::Bool(v) => Some(quote!(bool = #v)),
                    naga::Literal::I64(v) => Some(quote!(i64 = #v)),
                    naga::Literal::AbstractInt(v) => Some(quote!(i64 = #v)),
                    naga::Literal::AbstractFloat(v) => Some(quote!(f64 = #v)),
                },
                _ => None,
            }?;

            Some(quote!( pub const #name: #type_and_value;))
        })
        .collect()
}

#[cfg(test)]
mod tests {
    use super::*;

    use crate::assert_tokens_eq;
    use indoc::indoc;

    #[test]
    fn write_global_consts() {
        let source = indoc! {r#"
            const INT_CONST = 12;
            const UNSIGNED_CONST = 34u;
            const FLOAT_CONST = 0.1;
            // TODO: Naga doesn't implement f16, even though it's in the WGSL spec
            // const SMALL_FLOAT_CONST:f16 = 0.1h;
            const BOOL_CONST = true;

            @fragment
            fn main() {
                // TODO: This is valid WGSL syntax, but naga doesn't support it apparently.
                // const C_INNER = 456;
            }
        "#};

        let module = naga::front::wgsl::parse_str(source).unwrap();

        let consts = consts(&module);
        let actual = quote!(#(#consts)*);
        eprintln!("{actual}");

        assert_tokens_eq!(
            quote! {
                pub const INT_CONST: i32 = 12i32;
                pub const UNSIGNED_CONST: u32 = 34u32;
                pub const FLOAT_CONST: f32 = 0.1f32;
                pub const BOOL_CONST: bool = true;
            },
            actual
        );
    }
}
