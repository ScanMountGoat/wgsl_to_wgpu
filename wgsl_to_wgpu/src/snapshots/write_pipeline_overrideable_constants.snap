---
source: wgsl_to_wgpu/src/consts.rs
---
pub struct OverrideConstants {
    pub b1: Option<bool>,
    pub b2: Option<bool>,
    pub b3: bool,
    pub f1: Option<f32>,
    pub f2: f32,
    pub i1: Option<i32>,
    pub i2: i32,
    pub i3: Option<i32>,
    pub a: Option<f32>,
    pub b: Option<f32>,
}
impl OverrideConstants {
    pub fn constants(&self) -> Vec<(&'static str, f64)> {
        let mut entries = vec![
            ("b3", if self.b3 { 1.0 } else { 0.0 }),
            ("f2", self.f2 as f64),
            ("i2", self.i2 as f64),
        ];
        if let Some(value) = self.b1 {
            entries.push(("b1", if value { 1.0 } else { 0.0 }));
        };
        if let Some(value) = self.b2 {
            entries.push(("b2", if value { 1.0 } else { 0.0 }));
        };
        if let Some(value) = self.f1 {
            entries.push(("f1", value as f64));
        };
        if let Some(value) = self.i1 {
            entries.push(("i1", value as f64));
        };
        if let Some(value) = self.i3 {
            entries.push(("i3", value as f64));
        };
        if let Some(value) = self.a {
            entries.push(("0", value as f64));
        };
        if let Some(value) = self.b {
            entries.push(("35", value as f64));
        }
        entries
    }
}
